keps/sig-node/2403-pod-resources-allocatable-resources/README.md

```diff
@@ -0,0 +1,282 @@
+// WatchPodResourcesRequest is the request made to the Watch PodResourcesLister service
+message WatchPodResourcesRequest {}
+
+enum WatchPodAction {
+    ADDED = 0;
+    DELETED = 1;
+}
+
+// WatchPodResourcesResponse is the response returned by Watch function
+message WatchPodResourcesResponse {
+    WatchPodAction action = 1;
+    string uid = 2;
+    repeated PodResources pod_resources = 3;
+}
+
+// PodResources contains information about the node resources assigned to a pod
+message PodResources {
+    string name = 1;
+    string namespace = 2;
+    repeated ContainerResources containers = 3;
+}
+
+// ContainerResources contains information about the resources assigned to a container
+message ContainerResources {
+    string name = 1;
+    repeated ContainerDevices devices = 2;
+    repeated int64 cpu_ids = 3;
+}
+
+// Topology describes hardware topology of the resource
+message TopologyInfo {
+	repeated NUMANode nodes = 1;
+}
+
+// NUMA representation of NUMA node
+message NUMANode {
+	int64 ID = 1;
+}
+
+// ContainerDevices contains information about the devices assigned to a container
+message ContainerDevices {
+    string resource_name = 1;
+    repeated string device_ids = 2;
+    TopologyInfo topology = 3;
+}
+```
+
+### Test Plan
+
+The implementation PR adds a suite of E2E tests which cover both the existing `List` endpoint already implemented in the podresources API and
+the new proposed `GetAllocatableResources` API.
+
+### Graduation Criteria
+
+#### Alpha
+- [X] Implement the new service API.
+- [X] [Ensure proper e2e node tests are in place](https://k8s-testgrid.appspot.com/sig-node-kubelet#node-kubelet-serial&include-filter-by-regex=DevicePluginProbe).
+
+#### Alpha to Beta Graduation
+- [X] Demonstrate that the endpoint can be used to replace in-tree GPU device metrics in production environments (NVIDIA, sig-node April 30, 2019).
+
+#### Beta to G.A Graduation
+- [X] Multiple real world examples ([Multus CNI](https://github.com/intel/multus-cni)).
+- [X] Allowing time for feedback (2 years).
+- [X] [Start Deprecation of Accelerator metrics in kubelet](https://github.com/kubernetes/kubernetes/pull/91930).
+- [X] Risks have been addressed.
+
+### Upgrade / Downgrade Strategy
+
+With gRPC the version is part of the service name.
+Old versions and new versions should always be served and listened by the kubelet.
+
+To a cluster admin upgrading to the newest API version, means upgrading Kubernetes to a newer version as well as upgrading the monitoring component.
+
+To a vendor changes in the API should always be backwards compatible.
+
+### Version Skew Strategy
+
+Kubelet will always be backwards compatible, so going forward existing plugins are not expected to break.
+
+## Production Readiness Review Questionnaire
+### Feature enablement and rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Feature gate (also fill in values in `kep.yaml`).
+    - Feature gate name: `KubeletPodResources`.
```

Comment:
this is a feature gate that is already GA, we need to determine the best path for additive changes to existing GRPC endpoints.

Reply:
I know you put this on the sig-arch agenda, but the next meeting is post enhancements freeze. So, based on our slack discussion I would think this should get a new feature gate.

Reply:
Thanks @johnbelamaric for clarifying. I have no problems adding a feature gate, already in the process of changing the PR and I'm going to edit the KEP ASAP as well.

Followup question: we plan to add more extra APIs in the 1.22 cycle. Like the one we are proposing here, they will be completely independent and opt-in. Does this require a new featuregate or can we reuse the one I'm adding here?

In general, it seems to me we are inclined to add feature gate for each set of features we add to GA gRPC APIs, is my understanding correct?

Reply:
if they use the same feature gate, they will have to go through alpha/beta/stable at the same time. if that works for you, a single feature gate is OK.

Reply:
understood. We will evaluate in the 1.22+ cycle when we will propose the new changes. Updating this KEP as suggested. Thanks!